### ISSUE TITLE
Should Checkpoint save history as well?

### ISSUE THREAD

**Issue by ottonemo** [dnouri/skorch#289]
Continuation of the discussion in #246.

I think the points can be summarized to

1. should there be a `save_history` parameter
2. should we support pickling the whole model instead of using `save_*` methods (which consequently includes the history)
3. should we introduce a second `target` parameter for the history, e.g. `history_target` and `params_target`
4. should we find a better name for `target`

I think that saving the history alongside the model is a good idea, so yes to (1). Having the additional ability to use pickle for checkpointing via (2) sounds like a good idea but I'm not yet convinced where this feature makes life easier for the user, it would be nice to have an usage example that demonstrates the benefit of this feature. For (3) I would argue that we can re-use the existing parameter, e.g. strip the folder (if any) and append a static file name or append a `.history` suffix to the whatever is stored in `target`. Regarding (4) I think that something akin to `path` is a better choice but I don't have a strong opinion here.

**Comment by benjamin-work** [comment]
1) Then it's decided
2) First it can be a matter of convenience. Second, the code that loads the model may not care for whether it's a skorch NeuralNet or a completely different sklearn model. In that case, pickle would work but `save/load_params` would be useless.
3) I agree we could do it like that but this is a little bit less flexible.
4) We could try to be consistent with our existing code, i.e. use `f`. That name is not great in itself, but conforms to PyTorch.

**Comment by thomasjpfan** [comment]
1, 3, 4. `history_target` can be used define a path and configure `Checkpoint` to save the history. `params_target` is a good name in the context of `Checkpoint` since it is saving model parameters.

2. I think pickling the whole `NeuralNet ` object would make the model less portable. Current, after a user is done training a model in `skorch`, they easily move their module and parameters to another environment.

**Comment by benjamin-work** [comment]
I'm not too keen on the `*_target` names, is there a precedent for this?

**Comment by ottonemo** [comment]
It is not necessarily a path but possibly a file-like object. It denotes in every case the thing data is written to, i.e. the target destination of the checkpoint callback.

**Comment by benjamin-work** [comment]
Yes, it's a path or file-like, but PyTorch (and, as an extension, `skorch.NeuralNet.save_params`) calls this variable `f`. Granted, `f` is not the best variable name (though there is the convention of `with open(...) as f:`), but neither is target. When in doubt, I would opt for consistency.

**Comment by thomasjpfan** [comment]
Consider the following API:

```python
class Checkpoint(Callback):
    ...
    def __init__(
            self,
            f='',
            target_f='params.pt',
            history_f='history.json',
            monitor='valid_loss_best',
            sink=print,
    ):
```

`f` denotes the file of the pickled `NeuralNet`. By default, it is not pickled. What do you think?

**Comment by taketwo** [comment]
@ottonemo  your original suggestion

> For (3) I would argue that we can re-use the existing parameter, e.g. strip the folder (if any) and append a static file name or append a .history suffix to the whatever is stored in target.

does not account for the use case that you brought up later:

> It is not necessarily a path but possibly a file-like object. It denotes in every case the thing data is written to, i.e. the target destination of the checkpoint callback.

If the target argument (or whatever its name will be) can be a file-like object, then we can not automatically derive a name for the history file from it.

Thus unless we are willing to drop support for file-like objects, it seems we have to have three separate arguments as Thomas proposed.

**Comment by ottonemo** [comment]
What about 

```python
class Checkpoint(Callback):
    ...
    def __init__(
            self,
            target_f='params.pt',
            pickle_f=None,
            history_f=None,
            monitor='valid_loss_best',
            sink=print,
    ):
```

In that case we would support saving all three together and setting one of the parameters to `None` disables that functionality. This enables users to store all three formats separately (for whatever reason) but gives the flexibility to choose the format.

**Comment by BenjaminBossan** [comment]
Even though it's not super elegant, this is probably the most pragmatic solution. I agree that `None` is more explicit in meaning "don't save this" than `''`, though the latter should also not save.

Regarding the naming, I slightly prefer `f_*` instead of `*_f` because if sorted, similar arguments are next to one another. And I still don't like `target`, why not `params` like in `save_params`?

**Comment by taketwo** [comment]
Same here, I think `params` is more consistent:

```python
class Checkpoint(Callback):
    ...
    def __init__(
            self,
            f_params='params.pt',
            f_history=None,
            f_pickle=None,
            monitor='valid_loss_best',
            sink=noop,
    ):
```

This indeed is not too elegant, and saving both params/history to a custom directory (my use-case) is more verbose than I'd like it to be:

```python
callbacks.append(Checkpoint(f_params=(logdir / "model.pt"), f_history=(logdir / "history.json"))
```

This becomes even longer if template names are used. Of course, we can add one more argument for base directory and automatically prepend it when `f_*` parameters are relative paths. And also accept `f_*` parameters to be `True`, in which case a default filename is used. Then something like this will be possible:

```python
callbacks.append(Checkpoint(f_params=True, f_history=True, base_dir=logdir)
```

However I'd totally understand if you call this is overengineering :)

**Comment by benjamin-work** [comment]
> is more verbose than I'd like it to be

Tbh, I don't even think this is too verbose.

> However I'd totally understand if you call this is overengineering :)

I would really not introduce yet another parameter.

There is another way to do this that only requires one parameter, I'm putting it out here for discussion. What about only having one parameter that is a callable that receives the `net` instance as argument (a callback in a callback). Then you could do things like:

```
def on_checkpoint(net, logdir='./'):
    net.save_params(logdir + 'model.pt')
    net.save_history(logdir + 'model.history')
    with open(logdir + 'model.pkl') as f:
        pickle.dump(net, f)
```

This suggestion may even cover use cases that are not related to saving parameters. We could provide helper functions for the most common use case. But it is more complex than the other solutions.

**Comment by ottonemo** [comment]
> This suggestion may even cover use cases that are not related to saving parameters. We could provide helper functions for the most common use case. But it is more complex than the other solutions.

It is too complex for my taste. The result would be the that the user needs to plug together a checkpoint callback from utility functions which sounds nonintuitive to me. In that case it would probably better to have a checkpoint base class and to create subclasses for the most common use cases. That'd be more idiomatic wrt python as well.

But in the end I'm not sure whether this is worth it at all.

**Comment by taketwo** [comment]
@ottonemo  your last idea sounds in line with what you wrote before:

> In any case we should refactor `Checkpoint` in a way that makes extending its functionality easier (i.e. if I want to support my own protocol I should probably just need to inherit from `Checkpoint` and override one or two short functions).

So what's the conclusion?

**Comment by ottonemo** [comment]
In my opinion we should adopt the three parameter solution while refactoring the model saving part into separate methods so that it is easy to extend the checkpoint callback if necessary.

**Comment by ottonemo** [comment]
Resolved by merging #296.
